test/oracles/ChainlinkPriceOracle.sol

```diff
@@ -0,0 +1,148 @@
+} from "src/oracles/ChainlinkPriceOracle.sol";
+
+import {Utils} from "test/libraries/Utils.sol";
+
+contract ChainlinkPriceOracleTest is Test {
+    address private USDC = Utils.addressFromString("USDC");
+    address private WETH = Utils.addressFromString("WETH");
+    address private AMPL = Utils.addressFromString("AMPL");
+    address private USDCOracle = Utils.addressFromString("USDC Oracle");
+    address private WETHOracle = Utils.addressFromString("WETH Oracle");
+    address private AMPLOracle = Utils.addressFromString("AMPL Oracle");
+
+    ChainlinkPriceOracle internal oracle;
+
+    uint80 unusedRoundId = 69;
+    uint256 unusedStartedAt = 420;
+    uint256 updatedAt = 31336;
+    uint80 unusedAnsweredInRound = 70;
+
+    function setUp() public {
+        oracle = new ChainlinkPriceOracle();
+        vm.warp(31337);
+
+        vm.mockCall(USDCOracle, abi.encodeCall(AggregatorV3Interface.decimals, ()), abi.encode(uint8(8)));
+        vm.mockCall(WETHOracle, abi.encodeCall(AggregatorV3Interface.decimals, ()), abi.encode(uint8(8)));
+        vm.mockCall(AMPLOracle, abi.encodeCall(AggregatorV3Interface.decimals, ()), abi.encode(uint8(18)));
+        vm.mockCall(
+            USDCOracle,
+            abi.encodeCall(AggregatorV3Interface.latestRoundData, ()),
+            abi.encode(
+                unusedRoundId,
+                int256(1e8), // 1.00 with 8 decimals
+                unusedStartedAt,
+                updatedAt,
+                unusedAnsweredInRound
+            )
+        );
+        vm.mockCall(
+            WETHOracle,
+            abi.encodeCall(AggregatorV3Interface.latestRoundData, ()),
+            abi.encode(unusedRoundId, int256(1000e8), unusedStartedAt, updatedAt, unusedAnsweredInRound)
+        );
+        vm.mockCall(
+            AMPLOracle,
+            abi.encodeCall(AggregatorV3Interface.latestRoundData, ()),
+            abi.encode(unusedRoundId, int256(1.1e18), unusedStartedAt, updatedAt, unusedAnsweredInRound)
+        );
+    }
+
+    function getDefaultOracleData() internal view returns (ChainlinkPriceOracle.Data memory data) {
+        data = ChainlinkPriceOracle.Data({
+            token0Feed: USDCOracle,
+            token1Feed: WETHOracle,
+            timeThreshold: 1 days,
+            backoff: 1 days
+        });
+    }
+
+    function testReturnsExpectedPrice() public {
+        (uint256 priceNumerator, uint256 priceDenominator) =
+            oracle.getPrice(USDC, WETH, abi.encode(getDefaultOracleData()));
+        assertEq(priceNumerator, 1000e8);
+        assertEq(priceDenominator, 1e8);
+    }
+
+    function testReturnsInvertedPrice() public {
+        (uint256 priceNumerator, uint256 priceDenominator) =
+            oracle.getPrice(WETH, USDC, abi.encode(ChainlinkPriceOracle.Data(WETHOracle, USDCOracle, 1 days, 1 days)));
+        assertEq(priceNumerator, 1e8);
+        assertEq(priceDenominator, 1000e8);
+    }
+
+    function testNormalizedDecimals() public {
```

Comment:
Can we maybe also test how it behaves when one oracle has decimals > 18 (I believe it should revert because of underflow)

Reply:
it revert with underflow, should this be explicitly check and revert?

Reply:
Yes, explicitly check, and given that this is a failure condition that will never work, the revert should be `OrderNotValid(string)` with an applicable message for the watch-tower.